### DISCUSSION TITLE
nll scan of POI and nuisance parameters

### DISCUSSION THREAD

**Discussion by diptaparna** [scikit-hep/pyhf#1889] [Q&A] [answered]
Can you please tell me how to obtain the plot of nll as a function of model parameters (POI and nuisance parameters)?
The x-axis of the plot will be a scan of the parameter values, and the y-axis will be the negative log likelihood for a particular value. I want to make such plot for POI and each of the nuisance parameters separately.

**Comment by alexander-held** [comment, marked as answer]
Hi @diptaparna, you can get (twice the) NLL returned from a fit via the `return_fitted_val` keyword argument in `pyhf.infer.mle.fit`. Performing a scan means setting the parameter over which you are scanning to constant (via the `fixed_params` argument) and to the specific value you are interested in for the scan (via the `init_pars` argument). You then repeat that procedure for various parameter values and can then plot the NLL vs the values at which you held the respective parameter fixed.

[`cabinetry`](https://cabinetry.readthedocs.io/) provides you an API for (currently only one-dimensional) likelihood scans: [`cabinetry.fit.scan`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.fit.scan), which returns you an object that you can feed to [`cabinetry.visualize.scan`](https://cabinetry.readthedocs.io/en/latest/api.html#cabinetry.visualize.scan) for visualization purposes. A minimal example is below:
```python
import pyhf
import cabinetry

model = pyhf.simplemodels.correlated_background(
    signal=[12.0, 11.0],
    bkg=[50.0, 52.0],
    bkg_up=[45.0, 57.0],
    bkg_down=[55.0, 47.0],
)
data = [67, 58] + model.config.auxdata

scan_results = cabinetry.fit.scan(model, data, "mu")
cabinetry.visualize.scan(scan_results)
```
which will create the figure `figures/scan_mu.pdf`. You have to specify the parameter over which you want to scan (`"mu"` in this case), and you can optionally also specify the parameter range (see the documentation).

You can find the current `cabinetry` implementation for likelihood scans here: [`cabinetry/fit/__init__.py#L540-L637`](https://github.com/scikit-hep/cabinetry/blob/fce72088b4a6345304bf8c2e489938d41087a253/src/cabinetry/fit/__init__.py#L540-L637). An example of a setup for a 2d likelihood scan is [in this gist](https://gist.github.com/alexander-held/26724f1aa641632270a41079a8beac9c).
